### ISSUE TITLE
Error: Not compiled with GPU support on Windows (nvcc installed)

### ISSUE THREAD

**Issue by GiBucci** [facebookresearch/pytorch3d#1062]
Hi.

I am having trouble installing pytorch3d library on windows. I know it can be tricky, but I have only this solution at the moment.

I am sorry in advance if this is a duplicate question, or something trivial, but I spent quite some time looking at the other issues and I didn't find a solution.

I followed the INSTALL.md guide, created a new conda environment, installed all the necessary libraries, and then installed from source with "pip install "git+https://github.com/facebookresearch/pytorch3d.git".
It isn't specified if this works for windows too, but I tried and it seems it's fine anyway.

I encountered some troubles in importing the _C library but I tried installing Visual Studio tools and I managed to make it work anyway (I tried many things so I can't exactly tell you which steps I took, sorry).

I was finally ready to launch some demo tutorials, but when running the "deform_source_mesh_to_target_mesh.py" exercise, I find the error in the title, namely "RuntimeError: Not compiled with GPU support." when calling the underlying _C functions.

I saw that it might be a CUDA compiler error, so I downloaded all the tools, like nvcc. Now it's recognized correctly by the terminal, but it is not working.
I also tried setting some environment variables, like CUDA_HOME, pointing at the nvcc directory, but either it is not working or I'm doing something wrong.

I didn't apply any patch to the headers yet, it was hard to find the instructions, but it seems that those changes are answering another range of problems that I didn't encounter (yet).

It would be very appreciated if you could suggest me something in order to try and solve this matter.
Thank you in advance for your kindness and your consideration.

Here you can find some details on my environment. Let me know if I can provide additional information.

OS: Windows 11
NVIDIA RTX 3060

- conda list output:
# Name                    Version                   Build  Channel
addict                    2.4.0              pyhd8ed1ab_2    conda-forge
anyio                     3.5.0            py37h03978a9_0    conda-forge
argon2-cffi               21.3.0             pyhd8ed1ab_0    conda-forge
argon2-cffi-bindings      21.2.0           py37hcc03f2d_1    conda-forge
attrs                     21.4.0             pyhd8ed1ab_0    conda-forge
babel                     2.9.1              pyh44b312d_0    conda-forge
backcall                  0.2.0              pyh9f0ad1d_0    conda-forge
backports                 1.0                        py_2    conda-forge
backports.functools_lru_cache 1.6.4              pyhd8ed1ab_0    conda-forge
blas                      1.0                         mkl
bleach                    4.1.0              pyhd8ed1ab_0    conda-forge
brotlipy                  0.7.0           py37hcc03f2d_1003    conda-forge
ca-certificates           2021.10.26           haa95532_2
certifi                   2021.10.8        py37haa95532_2
cffi                      1.15.0           py37hd8e9650_0    conda-forge
charset-normalizer        2.0.10             pyhd8ed1ab_0    conda-forge
chumpy                    0.70                     pypi_0    pypi
colorama                  0.4.4              pyh9f0ad1d_0    conda-forge
cryptography              36.0.1           py37h65266a2_0    conda-forge
cuda                      11.6.0               h15814fa_0    nvidia
cuda-cccl                 11.6.55              hd268e57_0    nvidia
cuda-command-line-tools   11.6.0               h15814fa_0    nvidia
cuda-compiler             11.6.0               h15814fa_0    nvidia
cuda-cudart               11.6.55              h5fb1900_0    nvidia
cuda-cuobjdump            11.6.55              h18403f9_0    nvidia
cuda-cupti                11.6.55              h6b03428_0    nvidia
cuda-cuxxfilt             11.6.55              h3680034_0    nvidia
cuda-libraries            11.6.0               h15814fa_0    nvidia
cuda-libraries-dev        11.6.0               h15814fa_0    nvidia
cuda-memcheck             11.6.55              hfd78843_0    nvidia
cuda-nvcc                 11.6.55              hc626a03_0    nvidia
cuda-nvdisasm             11.6.55              h1c0b1b8_0    nvidia
cuda-nvml-dev             11.6.55              h2bb381e_0    nvidia
cuda-nvprof               11.6.55              h55b0447_0    nvidia
cuda-nvprune              11.6.55              h7586f89_0    nvidia
cuda-nvrtc                11.6.55              h2c369c2_0    nvidia
cuda-nvtx                 11.6.55              h62e5748_0    nvidia
cuda-nvvp                 11.6.58              h3b79e4d_0    nvidia
cuda-runtime              11.6.0               h15814fa_0    nvidia
cuda-sanitizer-api        11.6.55              h123461f_0    nvidia
cuda-toolkit              11.6.0               h15814fa_0    nvidia
cuda-tools                11.6.0               h15814fa_0    nvidia
cuda-visual-tools         11.6.0               h15814fa_0    nvidia
cudatoolkit               11.3.1               h59b6b97_2
cycler                    0.11.0             pyhd8ed1ab_0    conda-forge
debugpy                   1.5.1            py37hf2a7229_0    conda-forge
decorator                 5.1.1              pyhd8ed1ab_0    conda-forge
defusedxml                0.7.1              pyhd8ed1ab_0    conda-forge
einops                    0.3.0                      py_0    conda-forge
entrypoints               0.3             pyhd8ed1ab_1003    conda-forge
flit-core                 3.6.0              pyhd8ed1ab_0    conda-forge
freetype                  2.10.4               h546665d_1    conda-forge
fvcore                    0.1.5.post20220119          pypi_0    pypi
h5py                      2.9.0                    pypi_0    pypi
icu                       58.2             vc14hc45fdbb_0  [vc14]  anaconda
idna                      3.3                pyhd8ed1ab_0    conda-forge
importlib-metadata        4.10.1           py37h03978a9_0    conda-forge
importlib_resources       5.4.0              pyhd8ed1ab_0    conda-forge
intel-openmp              2022.0.0          h57928b3_3663    conda-forge
iopath                    0.1.9                    pypi_0    pypi
ipykernel                 6.7.0            py37h4038f58_0    conda-forge
ipython                   7.29.0           py37hd4e2768_0
ipython_genutils          0.2.0                      py_1    conda-forge
ipywidgets                7.6.5              pyhd8ed1ab_0    conda-forge
jedi                      0.18.1           py37h03978a9_0    conda-forge
jinja2                    3.0.3              pyhd8ed1ab_0    conda-forge
joblib                    1.0.1              pyhd8ed1ab_0    conda-forge
jpeg                      9b               vc14h4d7706e_1  [vc14]  anaconda
json5                     0.9.5              pyh9f0ad1d_0    conda-forge
jsonschema                4.4.0              pyhd8ed1ab_0    conda-forge
jupyter                   1.0.0            py37h03978a9_7    conda-forge
jupyter_client            7.1.2              pyhd8ed1ab_0    conda-forge
jupyter_console           6.4.0              pyhd8ed1ab_0    conda-forge
jupyter_core              4.9.1            py37h03978a9_1    conda-forge
jupyter_server            1.13.4             pyhd8ed1ab_0    conda-forge
jupyterlab                3.2.8              pyhd8ed1ab_0    conda-forge
jupyterlab_pygments       0.1.2              pyh9f0ad1d_0    conda-forge
jupyterlab_server         2.10.3             pyhd8ed1ab_0    conda-forge
jupyterlab_widgets        1.0.2              pyhd8ed1ab_0    conda-forge
kiwisolver                1.3.2            py37h8c56517_1    conda-forge
libblas                   3.9.0              12_win64_mkl    conda-forge
libcblas                  3.9.0              12_win64_mkl    conda-forge
libcublas                 11.8.1.74            h62d394a_0    nvidia
libcufft                  10.7.0.55            hfce90f6_0    nvidia
libcurand                 10.2.9.55            h4e24775_0    nvidia
libcusolver               11.3.2.107           h32ab9de_0    nvidia
libcusparse               11.7.1.55            hdcbd7c1_0    nvidia
liblapack                 3.9.0              12_win64_mkl    conda-forge
libnpp                    11.6.0.55            h2ebef2d_0    nvidia
libnvjpeg                 11.6.0.55            h33e88a0_0    nvidia
libpng                    1.6.37               h1d00b33_2    conda-forge
libprotobuf               3.18.0               h7755175_1    conda-forge
libsodium                 1.0.18               h8d14728_1    conda-forge
libtiff                   4.2.0                hd0e1b90_0
libuv                     1.43.0               h8ffe710_0    conda-forge
libwebp                   1.2.2                h57928b3_0    conda-forge
libwebp-base              1.2.2                h8ffe710_1    conda-forge
lz4-c                     1.9.3                h8ffe710_1    conda-forge
m2w64-gcc-libgfortran     5.3.0                         6    conda-forge
m2w64-gcc-libs            5.3.0                         7    conda-forge
m2w64-gcc-libs-core       5.3.0                         7    conda-forge
m2w64-gmp                 6.1.0                         2    conda-forge
m2w64-libwinpthread-git   5.0.0.4634.697f757               2    conda-forge
markupsafe                2.0.1            py37hcc03f2d_1    conda-forge
matplotlib                3.4.2            py37h03978a9_0    conda-forge
matplotlib-base           3.4.2            py37h4a79c79_0    conda-forge
matplotlib-inline         0.1.3              pyhd8ed1ab_0    conda-forge
mistune                   0.8.4           py37hcc03f2d_1005    conda-forge
mkl                       2021.4.0           h0e2418a_729    conda-forge
mkl-service               2.4.0            py37h75fcce0_0    conda-forge
mkl_fft                   1.3.1            py37h1874230_1    conda-forge
mkl_random                1.2.2            py37h90003fb_0    conda-forge
msys2-conda-epoch         20160418                      1    conda-forge
nbclassic                 0.3.5              pyhd8ed1ab_0    conda-forge
nbclient                  0.5.10             pyhd8ed1ab_1    conda-forge
nbconvert                 6.4.0            py37h03978a9_0    conda-forge
nbformat                  5.1.3              pyhd8ed1ab_0    conda-forge
nest-asyncio              1.5.4              pyhd8ed1ab_0    conda-forge
ninja                     1.10.2               h2d74725_1    conda-forge
notebook                  6.4.8              pyha770c72_0    conda-forge
numpy                     1.20.3           py37ha4e8547_0
numpy-base                1.20.3           py37hc2deb75_0
olefile                   0.46               pyh9f0ad1d_1    conda-forge
open3d                    0.13.0                   py37_0    open3d-admin
openssl                   1.1.1m               h2bbff1b_0
packaging                 21.3               pyhd8ed1ab_0    conda-forge
pandas                    1.3.1            py37h90003fb_0    conda-forge
pandoc                    2.17.0.1             h8ffe710_0    conda-forge
pandocfilters             1.5.0              pyhd8ed1ab_0    conda-forge
parso                     0.8.3              pyhd8ed1ab_0    conda-forge
pickleshare               0.7.5                   py_1003    conda-forge
pillow                    8.4.0            py37hd45dc43_0
pip                       21.1.3             pyhd8ed1ab_0    conda-forge
plyfile                   0.7.4              pyhd8ed1ab_0    conda-forge
portalocker               2.3.2            py37h03978a9_1    conda-forge
prometheus_client         0.13.0             pyhd8ed1ab_0    conda-forge
prompt-toolkit            3.0.24             pyha770c72_0    conda-forge
prompt_toolkit            3.0.24               hd8ed1ab_0    conda-forge
protobuf                  3.18.0           py37hf2a7229_0    conda-forge
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pygments                  2.11.2             pyhd8ed1ab_0    conda-forge
pyopenssl                 21.0.0             pyhd8ed1ab_0    conda-forge
pyparsing                 3.0.7              pyhd8ed1ab_0    conda-forge
pyqt                      5.9.2            py37h6538335_4    conda-forge
pyrsistent                0.18.1           py37hcc03f2d_0    conda-forge
pysocks                   1.7.1            py37h03978a9_4    conda-forge
python                    3.7.6           cpython_h60c2a47_6    conda-forge
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python_abi                3.7                     2_cp37m    conda-forge
pytorch                   1.10.1          py3.7_cuda11.3_cudnn8_0    pytorch
pytorch-mutex             1.0                        cuda    pytorch
pytorch3d                 0.6.1                    pypi_0    pypi
pytz                      2021.3             pyhd8ed1ab_0    conda-forge
pywin32                   303              py37hcc03f2d_0    conda-forge
pywinpty                  2.0.1            py37h7f67f24_0    conda-forge
pyyaml                    5.1.2            py37hfa6e2cd_0    conda-forge
pyzmq                     22.3.0           py37hcce574b_1    conda-forge
qt                        5.9.7            vc14h73c81de_0  [vc14]  anaconda
qtconsole                 5.2.2              pyhd8ed1ab_1    conda-forge
qtconsole-base            5.2.2              pyhd8ed1ab_1    conda-forge
qtpy                      2.0.0              pyhd8ed1ab_0    conda-forge
requests                  2.27.1             pyhd8ed1ab_0    conda-forge
scikit-learn              1.0.2            py37hcabfae0_0    conda-forge
scipy                     1.7.3            py37hb6553fb_0    conda-forge
send2trash                1.8.0              pyhd8ed1ab_0    conda-forge
setuptools                59.8.0           py37h03978a9_0    conda-forge
sip                       4.19.8          py37h6538335_1000    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
smplx                     0.1.28                   pypi_0    pypi
sniffio                   1.2.0            py37h03978a9_2    conda-forge
sqlite                    3.37.0               h8ffe710_0    conda-forge
tabulate                  0.8.9              pyhd8ed1ab_0    conda-forge
tbb                       2021.5.0             h2d74725_0    conda-forge
tensorboardx              2.2                pyhd8ed1ab_0    conda-forge
termcolor                 1.1.0                    pypi_0    pypi
terminado                 0.12.1           py37h03978a9_1    conda-forge
testpath                  0.5.0              pyhd8ed1ab_0    conda-forge
threadpoolctl             3.0.0              pyh8a188c0_0    conda-forge
tk                        8.6.7            vc14hb68737d_1  [vc14]  anaconda
torchaudio                0.10.1               py37_cu113    pytorch
torchvision               0.11.2               py37_cu113    pytorch
tornado                   6.1              py37hcc03f2d_2    conda-forge
tqdm                      4.61.2             pyhd8ed1ab_1    conda-forge
traitlets                 5.1.1              pyhd8ed1ab_0    conda-forge
trimesh                   3.9.24             pyh6c4a22f_0    conda-forge
typing_extensions         4.0.1              pyha770c72_0    conda-forge
ucrt                      10.0.20348.0         h57928b3_0    conda-forge
urllib3                   1.26.8             pyhd8ed1ab_1    conda-forge
vc                        14.2                 hb210afc_6    conda-forge
vs2015_runtime            14.29.30037          h902a5da_6    conda-forge
wcwidth                   0.2.5              pyh9f0ad1d_2    conda-forge
webencodings              0.5.1                      py_1    conda-forge
websocket-client          1.2.3              pyhd8ed1ab_0    conda-forge
wheel                     0.37.1             pyhd8ed1ab_0    conda-forge
widgetsnbextension        3.5.2            py37h03978a9_1    conda-forge
win_inet_pton             1.1.0            py37h03978a9_3    conda-forge
winpty                    0.4.3                         4    conda-forge
xz                        5.2.5                h62dcd97_1    conda-forge
yacs                      0.1.8                    pypi_0    pypi
yaml                      0.1.7            vc14h4cb57cf_1  [vc14]  anaconda
zeromq                    4.3.4                h0e60522_1    conda-forge
zipp                      3.7.0              pyhd8ed1ab_0    conda-forge
zlib                      1.2.11           vc14h1cdd9ab_1  [vc14]  anaconda
zstd                      1.4.9                h6255e5f_0    conda-forge

- nvcc -V output:
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2021 NVIDIA Corporation
Built on Fri_Dec_17_18:28:54_Pacific_Standard_Time_2021
Cuda compilation tools, release 11.6, V11.6.55
Build cuda_11.6.r11.6/compiler.30794723_0

- import pytorch3d gives no error
- from pytorch3d import _C gives no error
- torch.cuda.is_available() returns True
- torch.cuda.get_device_name(0) returns correctly 'NVIDIA GeForce RTX 3060 Laptop GPU'

**Comment by bottler** [comment]
I would clone this repository locally so and try to `pip install .` so that it doesn't have to download each time and you can make changes. Your problem I think is that PyTorch3D has done a CPU build and that's not what you want. For whether to build the CUDA parts of pytorch3d, you can see the condition [here](https://github.com/facebookresearch/pytorch3d/blob/v0.6.1/setup.py#L49) in `setup.py`. It is `if (torch.cuda.is_available() and CUDA_HOME is not None) or force_cuda:`. 

You can add code to check your expectations before that line, e.g.:
```
assert CUDA_HOME is not None
assert torch.cuda.is_available()
```

**Comment by GiBucci** [comment]
Hi, 

Thank you very much for your time and your help.

I figured that as well, because probably I messed up some installation order.

Now I uninstalled pytorch3d, cloned the repo locally, and tried reinstalling. 

The new problem is that both by building from x64 Native Tools Command Prompt, and the pip install -e . , it tells me that it's not finding the compiler I guess. The error message is:

UserWarning: Error checking compiler version for cl: [WinError 2] System cannot find the specified file
      warnings.warn(f'Error checking compiler version for {compiler}: {error}')
    error: [WinError 2] System cannot find the specified file

I tried setting the cl.exe path inside the Path environment variable, but it seems it is not working.

I am currently looking for solutions about this, if you already have any hint please let me know.

Thank you again very much in any case, and have a nice day.

**Comment by bottler** [comment]
Can you check which components of visual studio you have installed? You only get the compiler if you ask for it.

**Comment by GiBucci** [comment]
I have some troubles in understanding if the compiler has been installed correctly, Visual Studio installer is not clear.

I tried creating and compiling a cpp script in Visual Studio and it seems everything is working fine.

In order to bypass this problem I also installed a separate compiler, namely mingw-w64 from msys2, added to the path environment variable, and tested.
GCC is working fine: 
gcc --version
gcc.exe (Rev8, Built by MSYS2 project) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Both attempts did not solve the problem though, with the setup process returning the cl not found error.

What should I check after installing Visual Studio? How can I be sure the compiler can be seen also outside of visual studio?

**Comment by bottler** [comment]
Don't try to build with mingw - I'm pretty sure you won't be able to get it to work.

You must have the compiler. If you open the right VS code command prompt you should be able to type `cl`. Can you try each of the VS command prompt links in the start menu?

**Comment by GiBucci** [comment]
Ok then nevermind.

I can confirm you that if I type cl both in Anaconda Prompt and in x64 Native Tools Command Prompt this works just fine.

cl
Microsoft (R) C/C++ Optimizing Compiler versione 19.29.30139 per x86
Copyright (C) Microsoft Corporation. Tutti i diritti  sono riservati.

uso: cl [ opzione... ] nomefile... [ /link opzione di collegamento... ]

Even if the terminal shows this correctly, the installation (in both prompts) fails with the error in the EDIT above.
I'll report it here again for readability:

running build_ext

C:\Users\GiovanniBucci\anaconda3\envs\fit3d\lib\site-packages\setuptools\command\easy_install.py:159: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
EasyInstallDeprecationWarning,
C:\Users\GiovanniBucci\anaconda3\envs\fit3d\lib\site-packages\setuptools\command\install.py:37: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
setuptools.SetuptoolsDeprecationWarning,

**error: [WinError 2] Impossibile trovare il file specificato** (System cannot find the specified file)

Error in atexit._run_exitfuncs:
Traceback (most recent call last):
File "C:\Users\GiovanniBucci\anaconda3\envs\fit3d\lib\site-packages\colorama\ansitowin32.py", line 59, in closed
return stream.closed
ValueError: underlying buffer has been detached
ERROR: Command errored out with exit status 1: 'C:\Users\GiovanniBucci\anaconda3\envs\fit3d\python.exe' -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\Users\GiovanniBucci\Documents\WeWear\Model3D\pytorch3d\setup.py'"'"'; file='"'"'C:\Users\GiovanniBucci\Documents\WeWear\Model3D\pytorch3d\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(file) if os.path.exists(file) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, file, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.

Exception information:
Traceback (most recent call last):
File "C:\Users\GiovanniBucci\anaconda3\envs\fit3d\lib\site-packages\pip_internal\cli\base_command.py", line 180, in main
status = self.run(options, args)
File "C:\Users\GiovanniBucci\anaconda3\envs\fit3d\lib\site-packages\pip_internal\cli\req_command.py", line 205, in wrapper
return func(self, options, args)
File "C:\Users\GiovanniBucci\anaconda3\envs\fit3d\lib\site-packages\pip_internal\commands\install.py", line 402, in run
pycompile=options.compile,
File "C:\Users\GiovanniBucci\anaconda3\envs\fit3d\lib\site-packages\pip_internal\req_init.py", line 85, in install_given_reqs
pycompile=pycompile,
File "C:\Users\GiovanniBucci\anaconda3\envs\fit3d\lib\site-packages\pip_internal\req\req_install.py", line 773, in install
unpacked_source_directory=self.unpacked_source_directory,
File "C:\Users\GiovanniBucci\anaconda3\envs\fit3d\lib\site-packages\pip_internal\operations\install\editable_legacy.py", line 46, in install_editable
cwd=unpacked_source_directory,
File "C:\Users\GiovanniBucci\anaconda3\envs\fit3d\lib\site-packages\pip_internal\utils\subprocess.py", line 244, in call_subprocess
raise InstallationSubprocessError(proc.returncode, command_desc)
pip._internal.exceptions.InstallationSubprocessError: Command errored out with exit status 1: 'C:\Users\GiovanniBucci\anaconda3\envs\fit3d\python.exe' -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\Users\GiovanniBucci\Documents\WeWear\Model3D\pytorch3d\setup.py'"'"'; file='"'"'C:\Users\GiovanniBucci\Documents\WeWear\Model3D\pytorch3d\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(file) if os.path.exists(file) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, file, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.


It is still an error message about missing file, but it doesn't specify which is the missing file.
Last exit message talks about some kind of log but I don't understand where to find it and if that could be useful for this matter.

Any idea on what could it be? Same error anyway? Or different problem now?

**Comment by bottler** [comment]
When you say "the installation", what is the command you were running?

**Comment by GiBucci** [comment]
As you suggested, I made a local clone of the repository, and when I'm in the folder, I run either 

- python setup.py install
- pip install --verbose -e .

as stated by the install.md file or your first comment.

The result is the same with both commands, it stops at running build_ext as per my last comment.
I also tried removing the ./build folder, seems like nothing changes

**Comment by bottler** [comment]
Is there a file with build errors or build logs (something .txt) ?

**Comment by GiBucci** [comment]
I'd love to. 
In the error mentions something like a log file, but I can't find it anywhere.
If I run pip install with --log option specifying a file it outputs the error above without any additional information

I don't know if there's a log file on a lower level, and in case where it could be located. Still looking for it.

**Comment by GiBucci** [comment]
UPDATE 2:

I finally found the issue. 
CUDA_HOME was not inizialized corretly, as the path I inserted and the one expected by torch were overlapping by one folder.
Now it runs the compiler correctly.
Sadly more issues came out, mainly regarding compilation. Looking at other issues it seems like the compiler version is too new, I'll need to understand what should I change and how to downgrade the components.

If you are still willing to help me through the process I can keep you updated on the new errors, in other case thank you so much for your help until now and feel free to close the issue, since I think we are going off-topic.

Have a nice day.
